falcon/middlewares.py

```diff
@@ -1,16 +1,92 @@
+from typing import Iterable, Optional, Union
+
+from .request import Request
+from .response import Response
+
 
 class CORSMiddleware(object):
-    def process_response(self, req, resp, resource, req_succeeded):
-        """Implement a simple blanket CORS policy for all routes.
+    """CORS Middleware
+
+    This middleware provides a simple out-of-the box CORS policy, including handling
+    of preflighted requests from the browser.
+
+    See also:
+
+    * https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+    * https://www.w3.org/TR/cors/#resource-processing-model
+
+    Keyword Arguments:
+        allow_origins (Union[str, Iterable[str]]): List of origins to allow (case
+            sensitive). The string ``'*'`` acts as a wildcard, matching every origin.
+            (default ``'*'``).
+        expose_headers (Optional[Union[str, Iterable[str]]]): List of additional headers
+            to expose via the ``Access-Control-Expose-Headers`` header.
+            These headers are in addition to the CORS-safelisted ones:
+            ``Cache-Control``, ``Content-Language``, ``Content-Length``, ``Content-Type``,
+            ``Expires``, ``Last-Modified``, ``Pragma``. (default ``None``).
+
+            See also:
+            https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
+        allow_credentials (Optional[Union[bool, str, Iterable[str]]]): List of origins for
```

Comment:
Looking great now :+1: 

Tbh looking at this again I got another idea to un-union the interface :thinking: 

Could we possibly unify in the following way:
* `allow_origins` (unchanged, maybe we could add support for `None`?) -- list of origins, `*` means allow all
* `expose_headers` (unchanged) -- list of headers, `None` means expose none extra
* `allow_credentials` -- list of origins, `*` means allow all allowed by `allow_origins`, `None` means allow none

Thoughts?

Reply:
You know, I actually like this proposal more than the current one. 
The only thing we have to make clear is that `allow_credentials=='*'` only works for allowed origins, but that was the same as `True` before.

Reply:
I've updated the pr applying this suggestion